packages/repository/src/decorators/relation.decorator.ts

```diff
@@ -59,14 +69,42 @@ export function hasOne(definition?: Object) {
   return PropertyDecoratorFactory.createDecorator(RELATIONS_KEY, rel);
 }
 
-/**
- * Decorator for hasMany
- * @param definition
- * @returns {(target:any, key:string)}
- */
-export function hasMany(definition?: Object) {
-  const rel = Object.assign({type: RelationType.hasMany}, definition);
-  return PropertyDecoratorFactory.createDecorator(RELATIONS_KEY, rel);
+export function hasMany<T extends typeof Entity>(
+  definition?: Partial<HasManyDefinition>,
+  itemType?: T,
+): PropertyDecorator;
+
+export function hasMany<T extends typeof Entity>(
+  itemType: T,
+): PropertyDecorator;
+
+export function hasMany<T extends typeof Entity>(
+  definition?: Partial<HasManyDefinition> | T,
```

Comment:
With **current** implementation of hasMany (without a belongsTo counterpart), there is no need to infer the the target Model; just the foreign key (for the purpose of relations). As of right now, the parameters _required_ for the decorator should be `{keyFrom: someForeignKey}` and nothing else.

Once we have `belongsTo` however, we should be able to infer the foreign key just by getting the model constructor, but that may introduce the issue of circular dependency: e.g. Customer looks at Order to infer the foreign key and then Order looks at Customer to infer the primary key. But then again, TypeORM seem to do ok with it (see https://github.com/typeorm/typeorm#creating-a-many-to-one--one-to-many-relation), so maybe there isn't a problem here

For now, I think `Pick<Customer, fromKey>` should be the only required parameter of `@hasMany`

Reply:
I agree we may not need the target model in relation metadata, but we still do need it to correctly annotate the model property for things like JSON Schema generation.

Also as I commented below, a reasonable default for the foreign key can be generated using the same approach as we use in LB 3.x.

Maybe I am biased by my long LB 1.x-3.x experience, but I consider the following code snippet as the best user experience I'd like to have as an app developer:

```ts
class Customer extends Entity {
  // ...

  @hasMany(Order)
  orders: Order[]
}
```

Under the hood, I'd expect such annotation to:
 - Annotate `orders` as a property of `Order[]` type for the REST layer.
 - Configure the relation foreign key to `customerId`
 - Set any other relation configuration to a reasonable default that works for 80% of users

Reply:
Could you show what the `Order` class would look like in this case? Would users be expected to manually define `customerId` (I'm considering that this is a yes since #1475)?

Now that I think about it, maybe we don't need belongsTo metadata just yet to automatically configure `customerId`. We'd just need to look for what's been decorated as primary key in `Order` and call it a day. I'm assuming we should explicitly throw when we detect composite keys?

I think we'll include this change you're proposing in the next iteration of commit

Reply:
> Could you show what the Order class would look like in this case? Would users be expected to manually define customerId (I'm considering that this is a yes since #1475)?

It depends. If our decorators and repositories trigger the code in legacy juggler that handles relations, then the `customerId` property on `Order` class would be defined by juggler. It won't be exposed in TypeScript and won't be shown in JSON responses, but at least the persistence side will keep working. I am not sure if all different pieces of the puzzle will fit together that way though. (Think about different PersistedModel classes created by each repository at different time, etc.)

I think the `Order` class has to explicitly define the `customerId` property to make it accessible from TypeScript code and more importantly to tell the REST layer and `.toJSON()` method about it.